src/Main.hs

```diff
@@ -176,50 +176,152 @@ defaultStore =
+              Branch pn WildExp p -> (a, Branch pn (ENot nowhere (andRaw a)) p)
+              e@(Leaf _ c _) -> (c : a, e)
+              e@(Branch _ c _) -> (c : a, e)
+          )
+          []
+
+    -- flatten case tree
+    flatten :: [Exp Bool] -> [Exp Bool] -> [Exp Bool] -> Case Expr Post -> Err [Behaviour]
+    flatten iff postc pathcond (Leaf _ cond post) = do
+      c <- checkBool env cond
+      (p, maybeReturn, contracts) <- checkPost env contract post
+      return $ splitCase name False contract iface (c : pathcond) iff maybeReturn p postc contracts
+    flatten iff postc pathcond (Branch _ cond cs) = do
+      c <- checkBool env cond
+      leaves <- mapM (flatten iff postc (c : pathcond)) (normalize cs)
+      return $ join leaves
+
+splitBehaviour store (Constructor name contract iface@(Interface _ decls) iffs creates@(Creates assigns) extStorage maybeEnsures maybeInvariants) = do
+  let env = mkEnv contract store decls
+
+  rawUpdates <- mapM (checkAssign env) assigns
+  let stateUpdates = Map.fromList $ (\(id, upd) -> (id, Right <$> upd)) <$> rawUpdates
+
+  let calldataBounds = getCallDataBounds decls
+  iffs' <- checkIffs env (iffs <> calldataBounds)
+
+  invariants <- mapM (checkBool env) $ fromMaybe [] maybeInvariants
+  let storageBounds = fst $ getStorageBounds env
+      postcs = storageBounds <> invariants
+
+  return $ [(Invariants invariants [contract])]
+           ++ (splitCase name True contract iface [] iffs' Nothing stateUpdates postcs [contract])
+
+mkEnv :: Id -> Store -> [Decl]-> Env
+mkEnv contract store decls = (fromMaybe mempty (Map.lookup contract store), store, abiVars)
+ where
+   abiVars = Map.fromList $ map (\(Decl typ var) -> (var, metaType typ)) decls
+
+-- split case into pass and fail case
+splitCase :: String -> Bool -> String -> Interface -> [Exp Bool] -> [Exp Bool] -> Maybe ReturnExp
+          -> Map String [Either StorageLocation StorageUpdate] -> [Exp Bool] -> [String] -> [Behaviour]
+splitCase name creates contract iface ifs [] ret storage postcs contracts =
+  [ Behaviour name Pass creates contract iface (mconcat ifs) (mconcat postcs) contracts storage ret ]
+splitCase name creates contract iface ifs iffs ret storage postcs contracts =
+  [ Behaviour name Pass creates contract iface (mconcat (ifs <> iffs)) (mconcat postcs) contracts storage ret,
+    Behaviour name Fail creates contract iface (And (mconcat ifs) (Neg (mconcat iffs))) (mconcat postcs) contracts storage Nothing ]
+
+-- extracts bounds on Integer values in storage, returns Iff or Exp Bool
+-- representations for use in either pre or post conditions
+getStorageBounds :: Env -> ([Exp Bool], [IffH])
+getStorageBounds (ours, _, _) =
+  unzip $ catMaybes $ fmap getBound $ Map.toList ours
+  where
+    getBound :: (Id, SlotType) -> Maybe (Exp Bool, IffH)
+    getBound (id, (StorageValue typ)) = case metaType typ of
+      Integer -> Just $ (bound typ (IntVar id), IffIn nowhere typ [EntryExp nowhere id []])
```

Comment:
I'm not sure how to add bounds for non `Integer` types (e.g. `Bool`, `Bytestr`, mappings)? Would that even make sense?

Reply:
I don't think such bounds are necessary. We will be modeling those using `Bool` and `ByteString` as "metatypes", i.e. in the proving backends, so they won't need to be further constrained.